File: webapp/lib/grove/adapters/turns.ts
import type { SessionTurnView } from "@/lib/grove/api";

/**
 * Merging a windowed `/turns` response into the transcript already on screen.
 *
 * Pure, and separate from `transcript.ts` (which maps turns to assistant-ui
 * messages) because this answers a different question: not "how does a turn
 * render" but "may this window be placed against what we hold". That question
 * is where a transcript silently loses a turn, so it is the one worth pinning
 * without a daemon.
 */

/**
 * The fields of a turns response this merge reads.
 *
 * Deliberately NARROWER than `SessionDetailView` — the wire view is structurally
 * assignable to it, and naming only what the decision consumes keeps the merge
 * honest about its inputs rather than inheriting a shape it mostly ignores.
 */
export interface TurnWindow {
  turns: readonly SessionTurnView[];
  /** How many turns the session has in total, whatever this window carries. */
  total_turns: number;
  /** The index `turns[0]` sits at in the whole session. */
  first_turn_index: number;
  /** True only when the daemon HONOURED the cursor. False means "replace". */
  incremental: boolean;
}

/**
 * What `turnCursor` and `mergeTurns` need from what the client already holds
 * — just enough to place a new window against it, never the whole
 * `SessionDetailView` it happens to come from.
 */
export type HeldWindow = Pick<TurnWindow, "turns" | "first_turn_index">;

export type TurnMerge =
  /**
   * Use these turns as the whole transcript, starting at `first_turn_index`.
   * Carried explicitly rather than assumed to be `window.first_turn_index`,
   * because a follow-merge's result starts wherever the PREFIX it kept from
   * `held` starts, not wherever the fetched window itself started.
   */
  | { kind: "replace"; turns: readonly SessionTurnView[]; first_turn_index: number }
  /**
   * Nothing moved. Carries no turns ON PURPOSE: the caller must keep the array
   * it already has, and a payload here is how someone would accidentally hand
   * back a fresh-but-equal array — which re-runs the whole wire→assistant-ui
   * conversion for a transcript that did not change.
   */
  | { kind: "unchanged" }
  /** The window cannot be placed. Re-read the session with no cursor. */
  | { kind: "refetch" };

/**
 * The cursor to ask for, given what we already hold.
 *
 * ABSOLUTE — `held.first_turn_index + held.turns.length - 1` — never
 * `held.turns.length - 1` alone. The two coincide only while `held` starts at
 * session index 0, which was the ONLY shape a client could hold before
 * windowed reads existed; the moment the first read becomes a tail (see
 * `INITIAL_TURN_WINDOW`), `held.turns.length - 1` is a position *within the
 * window*, not the session, and asking for it as a cursor would silently
 * splice a hole between index 0 and wherever the window actually starts.
 *
 * Otherwise unchanged: `after_turn` is inclusive of its own index, turns are
 * append-*mostly*, and the last one keeps growing while the agent streams
 * parts and resolves tool calls. Asking past it would freeze a half-finished
 * turn on screen for the rest of the session. Holding nothing means no cursor
 * at all — there is no last-known turn to re-read.
 */
export function turnCursor(held: HeldWindow | undefined): number | undefined {
  return held && held.turns.length > 0 ? held.first_turn_index + held.turns.length - 1 : undefined;
}

/**
 * Place a window against the turns already held.
 *
 * Fail-safe by construction: anything this cannot prove contiguous asks for a
 * whole read rather than splicing a hole into the transcript. A cheap
 * transcript that has silently lost a turn is worse than an expensive one.
 *
 * The saving is on the WIRE, not in the render. Recombining is simply what a
 * windowed response requires; it does not make the client-side conversion
 * cheaper, because `messagesFromTurns` builds new message objects from every
 * turn whenever the array identity changes — and the array identity changes on
 * any tick that carries a tail. Keeping the prefix objects is still the right
 * default (it costs nothing and it is what a per-turn memo would need to become
 * useful), but do not mistake it for a live optimisation.
 *
 * TWO shapes of window reach here, and they take different branches on
 * purpose:
 *
 *   - An INCREMENTAL window is the answer to an `after_turn` cursor: the
 *     daemon has told us its start is exactly the index we asked to resume
 *     from, so it is provably contiguous with `held` and the two splice
 *     together, keeping `held`'s own prefix objects by reference.
 *   - A NON-incremental window carries no such promise — it is a `last`-sized
 *     fetch (including a "load earlier" widen) or the daemon declining a
 *     cursor it could not honour — so it is taken wholesale, as the complete
 *     answer in its own right. This is what makes a widen safe without any
 *     special-casing: `last` never claims contiguity with anything, and a
 *     widen's response already contains both the new earlier prefix AND the
 *     tail we already had, fetched together, so replacing wholesale is
 *     already "one contiguous range with no duplicated turns" — there is
 *     nothing left to splice.
 */
export function mergeTurns(held: HeldWindow | undefined, window: TurnWindow): TurnMerge {
  if (!window.incremental) {
    return { kind: "replace", turns: window.turns, first_turn_index: window.first_turn_index };
  }

  // Incremental against nothing to place it on. Only reachable if a cursor
  // went out for a transcript we no longer hold.
  if (!held) return { kind: "refetch" };

  const heldEnd = held.first_turn_index + held.turns.length;

  // A window starting past the end of what we hold would leave the turns
  // between unaccounted for. Contiguity is guaranteed by `turnCursor`, so this
  // means the request and the response disagree — never splice through it.
  if (window.first_turn_index > heldEnd) return { kind: "refetch" };

  // The offset into `held.turns` to keep, relative to where `held` itself
  // starts — NOT to `window.first_turn_index` alone, which was only ever a
  // valid slice point while `held` started at session index 0. A negative
  // offset means the window starts before `held` does, which `turnCursor`
  // never asks for; refuse rather than guess at what would fill the gap.
  const prefixCount = window.first_turn_index - held.first_turn_index;
  if (prefixCount < 0) return { kind: "refetch" };

  const turns = [...held.turns.slice(0, prefixCount), ...window.turns];
  if (turns.length === held.turns.length && window.turns.length === 0) return { kind: "unchanged" };
  return { kind: "replace", turns, first_turn_index: held.first_turn_index };
}

File: src/grove/daemon/_turns.py
"""How a session's turn list is windowed for one response.

Its own module because TWO routers need it — the authenticated
``/workspaces/{id}/sessions/{sid}/turns`` in ``app.py`` and the unauthenticated
``/public/{token}/turns`` in ``_public.py`` — and ``_public`` cannot import
``app`` without closing a cycle. One definition, so the public follower and the
private one cannot disagree about what a cursor means.
"""

from __future__ import annotations

from typing import NamedTuple

from grove.core.agents import SessionTurn


class TurnWindow(NamedTuple):
    """Which slice of a session's turns one response carries, and why."""

    turns: tuple[SessionTurn, ...]
    total: int
    first_index: int
    incremental: bool


def turn_window(
    turns: tuple[SessionTurn, ...],
    *,
    last: int | None,
    after_turn: int | None,
    before_turn: int | None = None,
) -> TurnWindow:
    """Resolve the requested window over a session's complete turn list.

    ``after_turn`` is INCLUSIVE of its own index, and that is the whole answer to
    the tail-mutation problem: turns are append-*mostly*, not append-only — the
    last turn keeps growing as the agent streams parts and resolves tool calls,
    while every earlier one is frozen (measured on a live session: 6 of 7 turns
    byte-identical over 45 s, only the tail moved). An exclusive cursor would
    freeze a half-finished turn on screen for the rest of the session, so the
    client's last-known turn is always re-sent and it replaces from
    ``first_index`` rather than blindly appending.

    A cursor STRICTLY BEYOND the end is the GAP: the session now holds fewer
    turns than the client claims to have seen, so the transcript was replaced
    or forked under the reader, ordinals no longer mean what the client thinks,
    and the honest answer is the whole session with ``incremental=False``.
    Fail-safe by construction — anything this cannot prove it can serve
    incrementally comes back whole, mirroring ``_SseHub.can_replay``'s fall
    back to a full snapshot.

    ``after_turn == total`` is deliberately NOT a gap but an empty incremental
    window: the client is exactly up to date, and answering a one-off-by-one
    cursor with the entire session would make the common "nothing happened"
    tick the most expensive request on the route.

    ``before_turn`` is the mirror of ``after_turn`` and the reason "load
    earlier" stops re-downloading the tail. It is EXCLUSIVE of its own index —
    the client already holds that turn, and unlike the forward case it is a
    FROZEN one (only the tail mutates), so re-sending it would buy nothing. With
    ``last`` it names a page size (``turns[n - last : n]``); without one it means
    every turn before ``n``. ``incremental`` stays False: this window is placed
    by its own ``first_turn_index`` against a prefix, not resumed from a cursor,
    and the client PREPENDS rather than replaces.

    A ``before_turn`` past the end CLAMPS to ``total`` rather than falling back
    to the whole session the way an over-run ``after_turn`` does, because the
    two make different claims. A forward cursor asserts "I have seen turn n", so
    a transcript shorter than that has been replaced under the reader. A
    backward one only asks for history preceding an endpoint, and an endpoint
    beyond the end is satisfied by everything there is.
    """
    total = len(turns)
    if before_turn is not None:
        end = min(before_turn, total)
        start = max(end - last, 0) if last is not None else 0
        return TurnWindow(turns[start:end], total, start, False)
    if after_turn is not None:
        if after_turn > total:
            return TurnWindow(turns, total, 0, False)
        return TurnWindow(turns[after_turn:], total, after_turn, True)
    if last is not None:
        start = max(total - last, 0)
        return TurnWindow(turns[start:], total, start, False)
    return TurnWindow(turns, total, 0, False)
